### ISSUE TITLE
Allow working with SemVers correctly

### ISSUE THREAD

**Issue by Tyil** [ugexe/zef#233]
Currently, I am using [semantic versioning](https://semver.org) for my projects. The [documentation on `META6.json`](https://docs.perl6.org/language/modules#index-entry-META6.json-META6.json) also promotes the idea of using major.minor.patch. This means backwards incompatible changes will bump the major version number.

## Context

I've been asking around for a while on how to achieve this idea in zef, to set a minimum version number on a dependency, and allow any higher version *except* a higher major version. I've recently gotten an answer to this question on https://github.com/scriptkitties/perl6-dist-helper/issues/34, and it seems to be unaivailable at this moment. To make good use of the benefits of semantic versioning, it is needed.

## Expected Behavior

A certain syntax in the META6.json to specify a minimum version. Zef will install a newer version if its available *unless* it's a higher major version.

## Actual Behavior

There is currently no such syntax available.

## Your Environment

* perl6 -v
```
This is Rakudo version 2018.01-143-g8ba3c86e7 built on MoarVM version 2018.01-77-g9a029b408
implementing Perl 6.c.
```

* zef list --installed
```
CORE:ver<6.c>:auth<perl>
App::Assixt:ver<0.1.3>
Bailador:ver<0.0.14>
Config::Parser::toml:ver<1.0.1>:auth<github:scriptkitties>
Config::TOML:ver<0.0.1>:auth<Andy Weidenbaum>
Config:ver<1.3.2>:auth<github:scriptkitties>
Crane:auth<Andy Weidenbaum>
DateTime::Parse:ver<0.0.1>
Digest::HMAC:ver<1.0.0>:auth<github:retupmoca>
Digest:ver<0.3.4>:auth<Lucien Grondin>
Dist::Helper:ver<0.19.1>:auth<github:scriptkitties>
Encode:ver<0.0.2>:auth<github:sergot>
File::Directory::Tree:auth<labster>
File::Find:ver<0.1>
File::Temp:ver<0.0.6>
File::Which:ver<1.0.0>
HTML::Entity:auth<Alexander Moquin>
HTML::Escape:ver<0.0.1>
HTML::Parser::XML:ver<0.1.2>:auth<github:tony-o>
HTTP::Easy:ver<1.1.0>
HTTP::HPACK:ver<0.9.2>
HTTP::MultiPartParser
HTTP::Request::Supply:ver<0.1.1>:auth<github:zostay>
HTTP::Server::Ogre:ver<0.0.3>
HTTP::Status
HTTP::UserAgent:ver<1.1.37>:auth<github:sergot>
Hash::Merge:ver<0.1.1>:auth<github:scriptkitties>
IO::Capture::Simple
IO::Socket::Async::SSL:ver<0.6.1>
IO::Socket::SSL:ver<0.0.1>:auth<github:sergot>
IRC::Client::Plugin::Github:ver<0.1.5>:auth<github:scriptkitties>
IRC::Client::Plugin::Ignore:ver<0.1.0>
IRC::Client::Plugin::NickServ:ver<0.1.1>
IRC::Client::Plugin::UrlTitle:ver<1.0.1>
IRC::Client::Plugin::UrlTitle:ver<1.1.0>
IRC::Client:ver<3.007006>
IRC::TextColor:ver<0.3.0>:auth<github:samcv>
JSON::Class:ver<0.0.10>:auth<github:jonathanstowe>
JSON::Fast:ver<0.9.11>
JSON::Fast:ver<0.9.6>
JSON::Marshal:ver<0.0.14>:auth<github:jonathanstowe>
JSON::Name:ver<0.0.2>:auth<github:jonathanstowe>
JSON::Unmarshal:ver<0.08>
LibraryMake:ver<1.0.0>:auth<github:retupmoca>
Linenoise:ver<0.1.1>:auth<Rob Hoelz>
Log::Any:ver<0.9.4>
META6:ver<0.0.19>:auth<github:jonathanstowe>
MIME::Base64:ver<1.2.1>:auth<github:retupmoca>
OpenSSL:ver<0.1.18>:auth<github:sergot>
PSGI:ver<1.2.0>
Path::Iterator:ver<0.0.1>:auth<Leon Timmermans>
SemVer:ver<0.1.3>
Shell::Command
TAP::Harness:ver<0.0.3>:auth<Leon Timmermans>
Template::Mojo:ver<.0.1>
Template::Mustache:ver<1.0.1>:auth<github:softmoth>
Term::termios:ver<0.0.1>:auth<github:titsuki>
Terminal::ANSIColor:ver<0.4>
Terminal::Getpass:ver<0.0.5>
Test::META:ver<0.0.13>:auth<github:jonathanstowe>
Test::Notice:ver<1.001003>
Test::Util::ServerPort:ver<0.0.1>:auth<github:jonathanstowe>
Test::When:ver<1.001008>
URI::Encode:ver<0.05>:auth<David Farrell>
URI:ver<.0.1.4>
URL::Find:ver<0.2.1>:auth<github:samcv>
XML:ver<0.0.3>:auth<Timothy Totten>
YAMLish:ver<0.0.4>:auth<Leon Timmermans>
zef:ver<0.2.1>:auth<github:ugexe>
zef:ver<0.2.2>:auth<github:ugexe>
```

**Comment by ugexe** [comment]
Semver only makes sense if every single module has to use it and it is built in to Rakudo. Anything not using semver would then not work like everything else ( letters and more/less than 3 parts are both currently perfectly valid in Rakudo ). If you “use Test:ver<6.c+>” then what? Invent a new superset of semver that allows letters, or only applies to versions with 3 parts? These would go against other parts of the spec regarding versions.

Specifying the floor of a range is needed, but I don’t see semver being the answer.

**Comment by Tyil** [comment]
I disagree with you here.

I'm not saying zef has to *only* allow SemVer practices anywhere, I want a syntax to be able to use it correctly. Something like `^1.2.3` to allow 1.2.3, up to (but not including) 2.0.0. For the modules that *do* use SemVer properly can actually utilize the benefits that come with it.

Not *every single module* needs to start using SemVer for something like this to be usable, but right now, SemVer is being promoted but it's advantages can't be used.

**Comment by ugexe** [comment]
I am interested in an implementation that is consistent with what exists and is used to determine module resolution in the core. There is no `use Foo:ver<^1.2.3>`. If things are not consistent then the resolution strategy chosen by zef may not reflect that of a given CUR. `perl6 -e 'Version.new($version-field-from-meta6-as-string) ~~ Version.new($version-field-from-meta6-as-string)'` must first do *all* of this. The user must be able to copy some dependency from `depends` and `use $that-exact-string`. This consistency between zef and CUR should not be underestimated.

In other words: figure out how to implement `^`-like functionality in perl6 version objects ( like how `+` and `*` ) in a backwards compatible way. Then zef will do exactly what you want, and `use Foo:ver<^1.2.3>` would resolve in a way consistent with zef.

**Comment by niner** [comment]
I think we actually already have everything we need to satisfy the original requirements. The role that semver assigns to the major version number my convention is fulfilled by the :api part of a modules long name. Different :api numbers explicitly tell the user that modules are incompatible. Thus by a module depending on Foo:api<1>:ver<1.2.3+> zef and rakudo are free to install or load higher versions as long as they provide the same :api. Thus we already have a ceiling. It's just not encoded into the version.

**Comment by ugexe** [comment]
@niner That could work. api still needs to be implemented on CUR though.

Here it is in CURI and CURFS from old PRs:
https://github.com/rakudo/rakudo/pull/1125/files#diff-9d9f3b6271234e410619d222c23d08f7R445
https://github.com/rakudo/rakudo/pull/1132/files#diff-d5dab2af6323f476ce331195a79ee4c8R157

Here are tests that demonstrates how this all works at the distribution and CUR level:
https://github.com/ugexe/zef2/blob/master/t/utils-distribution.t
https://github.com/ugexe/zef2/blob/master/t/compunit-repository-cache.t

Of course this is just my interpretation of `:api<>`

**Comment by niner** [comment]
Thanks @ugexe 
Despite much infrastructure already being there, there's a surprising number of places that still needed support for api. I've pushed it into the post release branch: https://github.com/rakudo/rakudo/commit/32c5c83c6e

**Comment by niner** [comment]
The difference to your PRs is that I do not treat :api as a Version. My gut says that simpler is better. Changing the api value really means we are not compatible anymore, so a straight forward match seems to fit the purpose of :api best. But that's a detail we can still adjust if better arguments surface.

**Comment by ugexe** [comment]
I lean towards it being a version so that highest version selection can take place on it. When does this matter one might ask? bin script wrappers are one instance. The `*` and `+` wouldn't likely be used, but would technically still work as a side effect of using a Version for its numerical sorting/comparison properties.

**Comment by ugexe** [comment]
As for why I state every module would need to use semver for it to work in a reasonable way: because we have `auth` -- something semver was not designed to accommodate. Consider: `Foo:ver<1.2.3>:auth<abc>` and `Foo:ver<1.x.3.z.6>:auth<xyz>`. Semver can only be applied to one of these, yet zef and CUR must make decisions that consider both of these for any given module query. Therefore any versioning format used should work the same for all modules.
